src/Control/Monad/Catch.hs

```diff
@@ -194,61 +217,97 @@ class MonadCatch m => MonadMask m where
+  -- result, in this case @_s2@, usually needs to be discarded, since those
+  -- effects have already been incorporated in the @release@ action.
+  --
+  -- The only effect which is intentionally not incorporated in the @release@
+  -- action is the effect of throwing an error. In that case, the error must be
+  -- re-thrown. One subtlety which is easy to miss is that in the case in which
+  -- @use@ and @release@ both throw an error, the error from @release@ should
+  -- take priority. Here is an implementation for @ExceptT@ which demonstrates
+  -- how to do this.
   --
   -- @
-  -- generalBracket acquire release cleanup use = ExceptT $
-  --   generalBracket
+  -- generalBracket acquire release use = ExceptT $ do
+  --   (eb, ec) <- generalBracket
   --     (runExceptT acquire)
-  --     (\eresource ->
-  --       case eresource of
-  --         Left _ -> return ()
-  --         Right resource -> runExceptT (release resource) >> return ())
-  --     (\eresource e ->
-  --        case eresource of
-  --          Left _ -> return ()
-  --          Right resource -> runExceptT (cleanup resource e) >> return ())
+  --     (\eresource exitCase -> case eresource of
+  --       Left e -> return (Left e) -- nothing to release, acquire didn't succeed
+  --       Right resource -> case exitCase of
+  --         ExitCaseSuccess (Right b) -> runExceptT (release resource (ExitCaseSuccess b))
+  --         ExitCaseException e       -> runExceptT (release resource (ExitCaseException e))
+  --         _                         -> runExceptT (release resource ExitCaseAbort))
   --     (either (return . Left) (runExceptT . use))
+  --   return $ do
+  --     -- The order in which we perform those two 'Either' effects determines
+  --     -- which error will win if they are both 'Left's. We want the error from
+  --     -- 'release' to win.
+  --     c <- ec
+  --     b <- eb
+  --     return (b, c)
   -- @
   --
-  -- In this implementation, we need to deal with the potential that
-  -- the @acquire@ action returned a @Left@ (as opposed to succeeding
-  -- with a @Right@ or throwing an exception via @throwM@), and
-  -- therefore have to handle the @Left@ case explicitly when provide
-  -- @release@, @cleanup@, and @use@ actions to the base monad's
-  -- implementation of @generalBracket@.
-  --
-  -- You should ensure that in all cases of the @acquire@ action
-  -- completing successfully, either the @release@ or @cleanup@
-  -- actions are called, regardless of what occurs in @use@.
-  --
   -- @since 0.9.0
   generalBracket
     :: m a
     -- ^ acquire some resource
-    -> (a -> m ignored1)
-    -- ^ release, no exception thrown
-    -> (a -> SomeException -> m ignored2)
-    -- ^ release, some exception thrown; the exception will be rethrown
+    -> (a -> ExitCase b -> m c)
+    -- ^ release the resource, observing the outcome of the inner action
     -> (a -> m b)
     -- ^ inner action to perform with the resource
-    -> m b
+    -> m (b, c)
```

Comment:
How do we feel about the order of type variables in `generalBracket`? Our primary motivator for picking this order of arguments was to match what `bracket` does, but now we have:

```haskell
bracket
  :: IO a
  -> (a -> IO b)
  -> (a -> IO c)
  -> IO c

generalBracket
  :: m a
  -> (a -> ExitCase b -> m c)
  -> (a -> m b)
  -> m (b, c)
```

Now, the type variables `b` and `c` are in the opposite order as in `bracket`. We could conceivably change this to:

```haskell
generalBracket
  :: m a
  -> (a -> ExitCase c -> m b)
  -> (a -> m c)
  -> m (b, c)
```

Or even:

```haskell
generalBracket
  :: m a
  -> (a -> ExitCase c -> m b)
  -> (a -> m c)
  -> m (c, b)
```

I don't have a strong opinion one way or the other. Thoughts?

Reply:
I don't have a strong opinion either. I'll change it to your first suggestion in order to at least match bracket's `(a -> m c)`, then change it back if someone feels strongly about this :)

Reply:
OK. The return type is interesting because it might hint at an order of evaluation. That is, the action which returns `c` runs before the action which returns `b`, hence `m (c, b)`. On the other hand, this puts us back into a state where the type variables are out-of-order again. Perhaps the order of evaluation is already obvious because `ExitCase c` appears as an argument to the function which returns an `m b`?

Reply:
Now that I am attempting the change, I feel like changing the variable names is a bad idea. It's important to thread the state in the correct order: from acquire to use, then from use to release. By swapping the order of `use` and `release` to match `bracket`, we have already made it less obvious that this is what implementors need to do. Let's at least keep these variable names as is in order to give a hint of the right ordering.

Reply:
Would it be crazy to change change the type variables in `bracket` to match `generalBracket`, to make it clearer that even though `use` is last so `bracket` curries better, it is executed in the middle of the `a` and `c` bits?

    bracket
     :: IO a
      -> (a -> IO c)
      -> (a -> IO b)
      -> IO b

Reply:
That sounds like a reasonable suggestion too (short of actually swapping the order of those arguments).

Reply:
done